File: python/magda/pipeline.py
from typing import Dict, Any, List
from .agents.operation_identifier import OperationIdentifier
from .agents.track_agent import TrackAgent
from .agents.clip_agent import ClipAgent
from .agents.volume_agent import VolumeAgent
from .agents.effect_agent import EffectAgent
from .agents.midi_agent import MidiAgent
from .agents.base import BaseAgent


class MAGDAPipeline:
    """Main pipeline orchestrator for MAGDA."""
    
    def __init__(self):
        self.operation_identifier = OperationIdentifier()
        self.agents: Dict[str, BaseAgent] = {
            "track": TrackAgent(),
            "clip": ClipAgent(),
            "volume": VolumeAgent(),
            "effect": EffectAgent(),
            "midi": MidiAgent(),
        }
    
    def process_prompt(self, prompt: str) -> Dict[str, Any]:
        """Process a natural language prompt through the MAGDA pipeline."""
        
        # Stage 1: Identify operations
        print("Stage 1: Identifying operations...")
        identification_result = self.operation_identifier.execute(prompt)
        operations = identification_result["operations"]
        
        print(f"Identified {len(operations)} operations:")
        for op in operations:
            op_type = op.get('type', 'unknown')
            op_desc = op.get('description', str(op))
            print(f"  - {op_type}: {op_desc}")
        
        # Stage 2: Execute operations with specialized agents
        print("\nStage 2: Executing operations...")
        results = []
        current_context = {}
        
        for operation in operations:
            op_type = operation["type"]
            op_description = operation["description"]
            op_parameters = operation.get("parameters", {})
            
            print(f"Processing {op_type} operation: {op_description}")
            
            # Find appropriate agent
            agent = self.agents.get(op_type)
            if not agent:
                print(f"Warning: No agent found for operation type '{op_type}'")
                continue
            
            # Execute operation
            try:
                # Combine operation description with parameters for context
                operation_text = f"{op_description}"
                if op_parameters:
                    operation_text += f" with parameters: {op_parameters}"
                
                result = agent.execute(operation_text, current_context)
                results.append(result)
                
                # Update context for next operations
                if op_type == "track" and "result" in result:
                    track_result = result["result"]
                    current_context["track_id"] = track_result.get("id")
                    current_context["track"] = track_result
                
                print(f"  ✓ {result['daw_command']}")
                
            except Exception as e:
                print(f"  ✗ Error executing {op_type} operation: {e}")
        
        return {
            "original_prompt": prompt,
            "operations": operations,
            "results": results,
            "daw_commands": [r["daw_command"] for r in results if "daw_command" in r]
        }
    
    def get_agent_info(self) -> Dict[str, Any]:
        """Get information about all available agents."""
        agent_info = {}
        for name, agent in self.agents.items():
            agent_info[name] = agent.get_agent_info()
        return agent_info 

File: cpp/src/pipeline.cpp
#include "magda_cpp/pipeline.hpp"
#include <memory>
#include <vector>
#include <algorithm>
#include <iostream>

namespace magda {

MAGDAPipeline::MAGDAPipeline(const std::string& api_key) : api_key_(api_key) {
    initializeAgents();
}

void MAGDAPipeline::initializeAgents() {
    // Initialize operation identifier
    operation_identifier_ = std::make_unique<OperationIdentifier>(api_key_);

    // Initialize specialized agents
    agents_["track"] = std::make_unique<TrackAgent>(api_key_);
    // agents_["clip"] = std::make_unique<ClipAgent>(api_key_);
    agents_["volume"] = std::make_unique<VolumeAgent>(api_key_);
    // agents_["effect"] = std::make_unique<EffectAgent>(api_key_);
    // agents_["midi"] = std::make_unique<MidiAgent>(api_key_);
}

std::optional<PipelineResult> MAGDAPipeline::processPrompt(const std::string& prompt) {
    try {
        std::vector<Operation> operations;
        std::vector<std::string> daw_commands;

        // Stage 1: Identify operations
        std::cout << "Stage 1: Identifying operations..." << std::endl;
        auto identification_result = operation_identifier_->identifyOperations(prompt);

        if (!identification_result.success) {
            std::cout << "Error identifying operations: " << identification_result.error_message << std::endl;
            return std::nullopt;
        }

        std::cout << "Identified " << identification_result.operations.size() << " operations:" << std::endl;
        for (const auto& op : identification_result.operations) {
            std::cout << "  - " << op.type << ": " << op.description << std::endl;
        }

        // Stage 2: Execute operations with specialized agents
        std::cout << "\nStage 2: Executing operations..." << std::endl;

        for (const auto& daw_operation : identification_result.operations) {
            std::cout << "Processing " << daw_operation.type << " operation: " << daw_operation.description << std::endl;

            // Convert DAWOperation to Operation
            OperationType op_type = OperationType::UNKNOWN;
            if (daw_operation.type == "track") op_type = OperationType::CREATE_TRACK;
            else if (daw_operation.type == "clip") op_type = OperationType::CREATE_CLIP;
            else if (daw_operation.type == "volume") op_type = OperationType::SET_VOLUME;
            else if (daw_operation.type == "effect") op_type = OperationType::ADD_EFFECT;
            else if (daw_operation.type == "midi") op_type = OperationType::CREATE_MIDI;

            Operation operation(op_type, daw_operation.parameters, daw_operation.type);

            // Find appropriate agent
            auto agent_it = agents_.find(daw_operation.type);
            if (agent_it == agents_.end()) {
                std::cout << "Warning: No agent found for operation type '" << daw_operation.type << "'" << std::endl;
                continue;
            }

            // Execute operation
            try {
                auto agent_response = agent_it->second->execute(daw_operation.description, context_);

                // Add to results
                operations.push_back(operation);
                daw_commands.push_back(agent_response.daw_command);

                // Update context with agent response
                if (!agent_response.context.is_null()) {
                    context_.merge_patch(agent_response.context);
                }

                std::cout << "  ✓ " << agent_response.daw_command << std::endl;

            } catch (const std::exception& e) {
                std::cout << "  ✗ Error executing " << daw_operation.type << " operation: " << e.what() << std::endl;
            }
        }

        return PipelineResult(operations, daw_commands, context_);

    } catch (const std::exception& e) {
        std::cout << "Pipeline error: " << e.what() << std::endl;
        return std::nullopt;
    }
}

BaseAgent* MAGDAPipeline::findAgentForOperation(const std::string& operation) {
    for (auto& [name, agent] : agents_) {
        if (agent->canHandle(operation)) {
            return agent.get();
        }
    }
    return nullptr;
}

} // namespace magda
